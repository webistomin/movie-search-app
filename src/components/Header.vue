<template>
  <header class="page-header">
    <div class="page-header__inner">
      <div class="page-header__block page-header__block--heading">
        <h1 class="page-header__logo">
          <router-link to="/" class="page-header__home">
            <span class="page-header__decor">Movie</span>searcher
          </router-link>
        </h1>
        <button class="page-header__toggler"
                :class="getToggleClass"
                @click="toggleAppNavigation"
                aria-label="Open navigation menu">
          <span></span><span></span><span></span>
        </button>
      </div>
      <div class="page-header__block page-header__block--search">
        <div class="page-header__wrapper">
          <input type="text" class="page-header__input" title="Search a movie" placeholder="Search a movie...">
          <svg class="page-header__icon page-header__icon--absolute" width="17" height="17">
            <use xlink:href="#icon-magnifier"></use>
          </svg>
        </div>
        <div class="page-header__buttons">
          <button class="page-header__button"
                  :class="{'page-header__button--active' : getRowViewState}"
                  @click="setRowView"
                  title="Row view">
            <svg class="page-header__icon" width="17" height="13">
              <use xlink:href="#icon-details"></use>
            </svg>
            Row view
          </button>
          <button class="page-header__button"
                  :class="{'page-header__button--active' : getGridViewState}"
                  @click="setGridView"
                  title="Grid view">
            <svg class="page-header__icon" width="17" height="13">
              <use xlink:href="#icon-grid"></use>
            </svg>
            Grid view
          </button>
        </div>
      </div>
      <div class="page-header__block page-header__block--user">
        <div class="page-header__account">
          <a href="#" class="page-header__btn btn" v-if="!getAuthorizeState">Sign in</a>
          <template v-else>
            <div class="page-header__userblock">
              <img src="/static/img/content/ava.jpg" width="32" height="32" alt="Avatar" class="page-header__avatar">
              <p class="page-header__text">Hello, <span class="page-header__name">Olia</span>!</p>
            </div>
            <div class="page-header__dropdown">
              <button class="page-header__button page-header__button--arrow"
                      :class="getUserButtonClass"
                      aria-label="Open user menu"
                      @click="toggleUserMenu">
                <svg class="page-header__icon page-header__icon--arrow" width="12" height="12">
                  <use xlink:href="#icon-arrow"></use>
                </svg>
              </button>
              <ul class="page-header__list" :class="getUserMenuClass">
                <li class="page-header__item">
                  <a href="#" class="page-header__link">
                    <svg class="page-header__icon page-header__icon--left" width="17" height="17">
                      <use xlink:href="#icon-heart"></use>
                    </svg>
                    Favorite list
                  </a>
                </li>
                <li class="page-header__item">
                  <a href="#" class="page-header__link">
                    <svg class="page-header__icon page-header__icon--left" width="17" height="17">
                      <use xlink:href="#icon-logout"></use>
                    </svg>
                    Log out
                  </a>
                </li>
              </ul>
            </div>
          </template>
        </div>
      </div>
    </div>
  </header>
</template>

<script>
  export default {
    name: 'Header',
    data() {
      return {
        isUserMenuOpened: false,
      };
    },
    methods: {
      setRowView() {
        this.$store.commit('setRowView', true);
        this.$store.commit('setGridView', false);
      },
      setGridView() {
        this.$store.commit('setGridView', true);
        this.$store.commit('setRowView', false);
      },
      toggleAppNavigation() {
        // eslint-disable-next-line no-unused-expressions
        this.getNavigationState ? this.$store.commit('setNavigationState', false) : this.$store.commit('setNavigationState', true);
      },
      toggleUserMenu() {
        this.isUserMenuOpened = !this.isUserMenuOpened;
      },
    },
    computed: {
      getUserMenuClass() {
        return this.isUserMenuOpened ? 'page-header__list--opened' : 'page-header__list--closed';
      },
      getUserButtonClass() {
        return this.isUserMenuOpened ? 'page-header__button--opened' : 'page-header__button--closed';
      },
      getAuthorizeState() {
        return this.$store.getters.getAuthorizeState;
      },
      getNavigationState() {
        return this.$store.getters.getNavigationState;
      },
      getRowViewState() {
        return this.$store.getters.getRowViewState;
      },
      getGridViewState() {
        return this.$store.getters.getGridViewState;
      },
      getToggleClass() {
        return this.getNavigationState ? 'page-header__toggler--opened' : 'page-header__toggler--closed';
      },
    },
  };
</script>

<style lang="sass">
  .page-header
    &__decor
      margin-right: 8px

    &__logo
      margin: 0

    &__home
      display: block
      position: relative
      font-family: $font-family--secondary
      font-size: 18px
      font-weight: 700
      text-transform: uppercase
      color: $color-text--secondary
      transition: all 0.3s ease-in-out
      outline: none

      &:hover,
      &:focus
        color: $color-gold
        transition: all 0.3s ease-in-out

      &:active
        color: $color-action
        transition: all 0.3s ease-in-out

      &::before
        width: 17px
        height: 6px
        content: ""
        position: absolute
        left: 5px
        top: 10px
        background-color: transparent
        background-image: url("~/static/img/content/scarf.svg")

    &__block
      display: flex
      align-items: center
      justify-content: space-between
      padding: 24px 12px

      &--heading,
      &--user,
      &--search
        background-color: $color-brown

    &__buttons
      display: none

    &__toggler
      font-size: 0
      width: 30px
      height: 20px
      background-color: transparent
      position: relative

      &:hover,
      &:focus
        span
          background-color: $color-gold
          transition: all 0.3s ease-in-out

      & span
        display: block
        width: inherit
        height: 2px
        background-color: $color-text--secondary
        position: absolute
        transition: all 0.3s ease-in-out

      &--closed
        & span
          &:first-child
            top: 0

          &:nth-child(2)
            top: 9px

          &:last-child
            bottom: 0

      &--opened
        & span
          &:first-child
            top: 8px
            transform: rotate(45deg)

          &:nth-child(2)
            top: 9px
            opacity: 0

          &:last-child
            transform: rotate(-45deg)
            bottom: 10px

    &__wrapper
      width: 100%
      position: relative

    &__icon
      transition: all 0.3s ease-in-out
      fill: $color-brown

      &--absolute
        position: absolute
        right: 10px
        top: 50%
        transform: translateY(-50%)
        fill: #d4d8da

      &--left
        fill: $color-text
        position: absolute
        left: calc(50% - 70px)
        transition: all 0.3s ease-in-out

      &--white
        fill: $color-text--secondary
        margin-right: 10px

      &--arrow
        fill: $color-text--secondary

    &__input
      width: 100%
      padding: 16.5px 10px
      padding-right: 30px
      box-sizing: border-box
      border: 1px solid #d4d8da
      border-radius: 5px
      font-weight: 400
      font-size: 13px
      background-color: transparent
      color: $color-text--secondary
      transition: all 0.3s ease-in-out

      &:hover
        border-color: $color-gold

        & + .page-header__icon--absolute
          fill: $color-gold

      &:focus
        border-color: $color-gold
        outline: none

        & + .page-header__icon--absolute
          fill: $color-gold

      &::placeholder
        color: $color-text--secondary

    &__account
      width: 100%
      display: flex
      justify-content: space-between
      align-items: center

    &__avatar
      border-radius: 50%
      margin-right: 16px
      border: 3px solid $color-gold

    &__button
      transition: all 0.3s ease-in-out

      &:hover,
      &:focus
        .page-header__icon
          transition: all 0.3s ease-in-out
          fill: $color-gold

      &--arrow
        width: 32px
        height: 32px
        font-size: 0
        outline: none

      &--closed
        transform: rotate(0deg)

      &--opened
        transform: rotate(-180deg)

    &__userblock
      display: flex
      align-items: center

    &__btn
      display: flex
      display: none
      align-items: center
      justify-content: center
      width: 100px
      text-align: center
      margin: 0 auto

    &__text
      font-size: 13px
      font-weight: 400
      color: $color-text--secondary
      position: relative
      max-width: 100px
      overflow: hidden
      text-overflow: ellipsis

    &__name
      text-decoration: underline

    &__list
      display: block
      position: absolute
      top: 250px
      left: 0
      width: 100%
      background-color: $color-brown--secondary
      z-index: 1

      &--closed
      transform: scaleY(0)
      transform-origin: top
      transition: transform 0.3s ease-in-out

      &--opened
        transform: scaleY(1)
        transform-origin: top
        transition: transform 0.3s ease-in-out

    &__link
      box-sizing: border-box
      position: relative
      padding: 15px 0
      display: flex
      align-items: center
      justify-content: center
      font-size: 13px
      font-weight: 400
      color: $color-text
      transition: all 0.3s ease-in-out
      outline: none

      &:hover,
      &:focus
        color: $color-text--secondary
        transition: all 0.3s ease-in-out

        & .page-header__icon
          fill: $color-text--secondary
          transition: all 0.3s ease-in-out

      &:active
        color: $color-action
        transition: all 0.3s ease-in-out

        & .page-header__icon
          fill: $color-action
          transition: all 0.3s ease-in-out


    @media (min-width: $screen-sm)

      &__inner
        display: flex
        flex-wrap: wrap

      &__block
        box-sizing: border-box

        &--heading
          width: 100%

        &--search,
        &--user
          width: 50%

      &__list
        top: 165px

    @media (min-width: $screen-md)
      &__decor
        display: none

      &__inner
        flex-wrap: nowrap
        align-items: flex-start

      &__block
        box-sizing: border-box
        flex-shrink: 0
        min-height: 60px

        &--heading,
        &--user
          width: 20vw

        &--search
          width: 60vw
          padding: 13px 12px
          flex-shrink: 0
          padding-left: 30px
          padding-right: 56px

        &--heading
          position: relative
          padding: 20px 12px
          padding-left: 45px

          &::after
            content: ""
            position: absolute
            top: 0
            right: 0
            width: 5px
            height: 100%
            background-color: rgba(255, 255, 255, 0.1)

        &--user
          padding: 11px 12px
          position: relative

          &::after
            content: ""
            position: absolute
            top: 0
            left: 0
            width: 5px
            height: 100%
            background-color: rgba(255, 255, 255, 0.1)

      &__toggler
        display: none

      &__input
        padding-top: 8.5px
        padding-bottom: 8.5px

      &__buttons
        display: flex
        margin-left: 42px

      &__button
        font-size: 0
        width: 34px
        height: 34px
        outline: none

        & .page-header__icon
          fill: $color-text--secondary

        &--active
          & .page-header__icon
            fill: $color-gold

      &__account
        justify-content: center

      &__list
        width: 20vw
        top: 60px
        left: 0
        z-index: 10
        background-color: $color-brown

    @media (min-width: $screen-lg)
      &__decor
        display: inline
</style>
