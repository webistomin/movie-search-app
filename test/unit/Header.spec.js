import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Header from '../../src/components/Header';
import routes from '../../src/router/';

describe('Header.vue', () => {
  let actions;
  let store;
  let state;
  let mutations;
  let getters;
  let router;
  let vm;
  let mockedRoute;
  let localVue;

  beforeEach(() => {
    localStorage.clear();
    jest.clearAllMocks();
    jest.resetModules();

    actions = {
      fetchMoviesWithSearchQuery: jest.fn().mockImplementation(() => Promise.resolve('called')),
      removeNewSession: jest.fn(),
    };

    mutations = {
      setRequestToken: (state, payload) => {
        state.requestToken = payload;
      },
      setRowView: (state, payload) => {
        state.isRowView = payload;
      },
      setGridView: (state, payload) => {
        state.isGridView = payload;
      },
      setNavigationState: (state, payload) => {
        state.isNavigationOpened = payload;
      },
    };

    state = {
      requestToken: null,
      isRowView: false,
      isGridView: true,
      isNavigationOpened: false,
    };

    getters = {
      getNavigationState: state => state.isNavigationOpened,
    };

    store = new Vuex.Store({
      state,
      mutations,
      actions,
      getters,
    });

    localVue = createLocalVue();
    localVue.use(Vuex);
    router = routes;
  });

  it('Matches snapshot', () => {
    vm = shallowMount(Header, {
      store,
      localVue,
      router,
    }).vm;
    expect(vm.$el).toMatchSnapshot();
  });

  it('Sets the correct default data', () => {
    vm = shallowMount(Header, {
      store,
      localVue,
      router,
    }).vm;
    expect(vm.$data.isUserMenuOpened).toEqual(false);
  });

  it('Has a created hook', () => {
    vm = shallowMount(Header, {
      store,
      localVue,
      router,
    }).vm;
    expect(typeof Header.created).toBe('function');
  });

  describe('test "fetchMoviesWithSearchQuery" method', () => {
    it('call "fetchMoviesWithSearchQuery" action', () => {
      vm = shallowMount(Header, {
        store,
        localVue,
        router,
      }).vm;
      vm.fetchMoviesWithSearchQuery();
      expect(actions.fetchMoviesWithSearchQuery).toHaveBeenCalled();
    });

    it('push "/search" to router', async () => {
      vm = shallowMount(Header, {
        store,
        localVue,
        router,
      }).vm;
      await vm.fetchMoviesWithSearchQuery();
      expect(vm.$route.path).toEqual('/search');
    });
  });

  describe('test "logOut" method', () => {
    it('call "removeNewSession" action', () => {
      vm = shallowMount(Header, {
        store,
        localVue,
        router,
      }).vm;
      vm.logOut();
      expect(actions.removeNewSession).toHaveBeenCalled();
    });
  });

  describe('test "setRowView" method', () => {
    it('commit "setRowView" and "setGridView"', () => {
      vm = shallowMount(Header, {
        store,
        localVue,
        router,
      }).vm;
      vm.setRowView();
      expect(state.isRowView).toEqual(true);
      expect(state.isGridView).toEqual(false);
    });
  });

  describe('test "setGridView" method', () => {
    it('commit "setRowView" and "setGridView"', () => {
      vm = shallowMount(Header, {
        store,
        localVue,
        router,
      }).vm;
      vm.setGridView();
      expect(state.isRowView).toEqual(false);
      expect(state.isGridView).toEqual(true);
    });
  });

  describe('test "toggleAppNavigation" method', () => {
    it('commit "setNavigationState" with false if navigation opened', () => {
      vm = shallowMount(Header, {
        store,
        localVue,
        router,
      }).vm;
      store.commit('setNavigationState', true);
      vm.toggleAppNavigation();
      expect(state.isNavigationOpened).toEqual(false);
    });

    it('commit "setNavigationState" with true if navigation closed', () => {
      vm = shallowMount(Header, {
        store,
        localVue,
        router,
      }).vm;
      store.commit('setNavigationState', false);
      vm.toggleAppNavigation();
      expect(state.isNavigationOpened).toEqual(true);
    });
  });
});
