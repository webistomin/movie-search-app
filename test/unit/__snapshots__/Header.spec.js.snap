// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header.vue Matches snapshot 1`] = `
<header
  class="page-header"
  role="banner"
>
  <div
    class="page-header__inner"
  >
    <div
      class="page-header__block page-header__block--heading"
    >
      <h1
        class="page-header__logo"
      >
        <router-link-stub
          class="page-header__home"
          event="click"
          tag="a"
          to="/"
        >
          <span
            class="page-header__decor"
          >
            Movie
          </span>
          searcher
        
        </router-link-stub>
      </h1>
       
      <button
        aria-label="Open navigation menu"
        class="page-header__toggler page-header__toggler--closed"
      >
        <span />
        <span />
        <span />
      </button>
    </div>
     
    <div
      class="page-header__block page-header__block--search"
    >
      <div
        class="page-header__wrapper"
      >
        <input
          class="page-header__input"
          placeholder="Search a movie..."
          title="Search a movie"
          type="text"
        />
         
        <svg
          class="page-header__icon page-header__icon--absolute"
          height="17"
          width="17"
        >
          <use
            xlink:href="#icon-magnifier"
          />
        </svg>
      </div>
       
      <div
        class="page-header__buttons"
      >
        <button
          class="page-header__button"
          title="Row view"
        >
          <svg
            class="page-header__icon"
            height="13"
            width="17"
          >
            <use
              xlink:href="#icon-details"
            />
          </svg>
          
          Row view
        
        </button>
         
        <button
          class="page-header__button"
          title="Grid view"
        >
          <svg
            class="page-header__icon"
            height="13"
            width="17"
          >
            <use
              xlink:href="#icon-grid"
            />
          </svg>
          
          Grid view
        
        </button>
      </div>
    </div>
     
    <div
      class="page-header__block page-header__block--user"
    >
      <div
        class="page-header__account"
      >
        <a
          class="page-header__btn btn"
          href="https://www.themoviedb.org/authenticate/undefined?redirect_to=http://localhost:8080/"
        >
          Log in
        </a>
      </div>
    </div>
  </div>
</header>
`;
